src/sqlite3gen.cpp

```diff
@@ -930,7 +930,7 @@ static int insertPath(QCString name, bool local=TRUE, bool found=TRUE, int type=
 static void recordMetadata()
 {
   bindTextParameter(meta_insert,":doxygen_version",getFullVersion());
-  bindTextParameter(meta_insert,":schema_version","0.2.1",TRUE); //TODO: this should be a constant somewhere; not sure where
+  bindTextParameter(meta_insert,":schema_version",QCString("0.2.1")); //TODO: this should be a constant somewhere; not sure where
```

Comment:
Why is `TRUE` gone here and in line 2440 as the default is `FALSE` for this parameter?

Reply:
The _short_ answer is just that I still see the same symptom (schema_version is junk/empty) with the TRUE present. (Here's a fresh CI job, https://github.com/abathur/doxy_db/actions/runs/4765677175/jobs/8471719090, which uses doxygen built from https://github.com/abathur/doxygen/commit/3d4906a0fbd2ebfef08c8ce89ab57ccad3c5dc73.)

The default for `_static` causes `bindTextParameter` to pass `SQLITE_TRANSIENT` to `sqlite3_bind_text`, which in turn tells sqlite to go ahead and make a copy of the data because the pointer could change.

Passing `TRUE` tells it to pass `SQLITE_STATIC`, which tells sqlite that it doesn't need to make a copy because the value won't change.

It seems plausible to me that changes in 592aaa4f17 rendered the `_static` argument moot (i.e., `SQLITE_TRANSIENT` may now always be correct), but I don't write much C++ and don't completely understand all of the underlying motives and implications of that refactor.

Reply:
If you're confident that it's correct, I'm happy to update the PR to remove the `_static` argument and update `bindTextParameter` to always pass `SQLITE_TRANSIENT` to `sqlite3_bind_text`.

Reply:
Thanks for picking at this.

I'll respond in a bit more detail later for posterity, but I was just parroting some conversions from `"blah"` -> `QCString("blah")` that I saw elsewhere in 592aaa4, and it looks like all I actually need here is to fix the _STATIC/_TRANSIENT bit. I'll force push to update soon.

Reply:
> It is a difficult problem and I also reer to the commits:
> 
> ```
> commit ee9a4679ae630f8f808619c5450f6b2f5a860994
> Author: Travis A. Everett <travis.a.everett@gmail.com>
> Date:   Tue May 5 09:07:15 2020 -0500
> 
>     sqlite: flip _static default in bindTextParameter
> 
> commit 2a6f9d50d606c59e86fe99e4304f056fd7f1032c
> Author: Travis A. Everett <travis.a.everett@gmail.com>
> Date:   Mon May 4 17:50:47 2020 -0500
> 
>     fix #7746 with liberal use of _static==FALSE
> 
>     Something in the refactor at 960bae7 caused junk data to start showing up in
>     some sqlite3 columns. Calling bindTextParameter with _static==FALSE causes it
>     to call sqlite3_bind_text with SQLITE_TRANSIENT instead of SQLITE_STATIC,
>     which in turn causes sqlite3 to copy the string at bind time.
> ```

For broader context, when I did the refactoring work on sqlite3gen that got merged in 2018, there were quire a few places where junk data could end up in some database fields. My confidence in C++ is minimal, so I was trying to be conservative and only tried to fix fields I had seen junk data in at some point myself. 

IIRC in 2020 I was fixing up some build failures that had crept in, and in the process I noticed more of these fields with junk data because I had been to conservative (see discussion in https://github.com/doxygen/doxygen/pull/7750). Initially that PR was just explicitly passing _static=FALSE to more of the calls and in https://github.com/doxygen/doxygen/pull/7750#issuecomment-624046822 you encouraged me to invert the default to minimize how many call sites override it.

>  * It is a bit strange to me that you need the explicit conversion to `QCString` as I thought this would be done automatically (by means of `QCString( const char *str ) : m_rep(str?str:"") {}`, taken from qcstring.h)

Good catch. I was basically just parroting some of the transformations that were already in 592aaa4f17, such as:
- https://github.com/doxygen/doxygen/commit/592aaa4f17d73ec8c475df0f44efaea8cc4d575c#diff-c9367a6d87716cac78bb0ce2a1018084931694345c1c27e61ffbc71a79531674L2864-R2867
and
- https://github.com/doxygen/doxygen/commit/592aaa4f17d73ec8c475df0f44efaea8cc4d575c#diff-c9367a6d87716cac78bb0ce2a1018084931694345c1c27e61ffbc71a79531674L3073-R3074

I was reading these notes in the commit as supporting that:
```
    - The implicit convert from 'QCString' to 'const char *' is removed
    - Strings parameters use 'const QCString &' as much as possible in favor
      over 'const char *'
```

But now that you call it out, I realize they say nothing about the inverse conversion.

> * do you see the problems with `_static` is `TRUE` and with `FALSE`?
>   This is a bit hard to get from the debug output you give as it is regarding `schema_version` and I expect that this would be one of the first things to check / retrieve.

Just with `TRUE`, I think. It's hitting on `schema_version` because the tool I'm updating ([doxy_db](https://github.com/abathur/doxy_db) asserts what schema version it's compatible with, so it's failing before it does much.

> might it be that the `SQLITE_STATIC` part has problems with reading the data?

My mental model for what it's doing is that when `SQLITE_STATIC` is used at bind time on something that could actually change before the statement is finally executed, it'll hold on to the memory location and then later insert whatever is currently memory where it expected the value to be. (These are wrong in the generated database.) Always using `SQLITE_TRANSIENT` with QCStrings at least appears to fix this (without explicitly converting to QCString) in testing on my end.